src/device_manager.rs

```diff
@@ -0,0 +1,431 @@
+    /// Range mapping for VM exit mmio operations.
+    mmio_bus: BTreeMap<Range, Arc<Mutex<dyn Device>>>,
+    /// Range mapping for VM exit pio operations.
+    pio_bus: BTreeMap<Range, Arc<Mutex<dyn Device>>>,
+}
+
+impl DeviceManager {
+    /// Create a new `DeviceManager` with a `SystemAllocator` reference which would be
+    /// used to allocate resource for devices.
+    pub fn new(resource: Arc<Mutex<SystemAllocator>>) -> Self {
+        DeviceManager {
+            resource,
+            devices: HashMap::new(),
+            mmio_bus: BTreeMap::new(),
+            pio_bus: BTreeMap::new(),
+        }
+    }
+
+    fn insert(&mut self, dev: DeviceDescriptor) -> Result<()> {
+        // Insert if the key is non-present, else report error.
+        if self.devices.get(&(dev.instance_id)).is_some() {
+            return Err(Error::Exist);
+        }
+        self.devices.insert(dev.instance_id, dev);
+        Ok(())
+    }
+
+    fn remove(&mut self, instance_id: u32) -> Option<DeviceDescriptor> {
+        self.devices.remove(&instance_id)
+    }
+
+    fn device_descriptor(
+        &self,
+        id: u32,
+        name: String,
+        dev: Arc<Mutex<dyn Device>>,
+        parent_bus: Option<Arc<Mutex<dyn Device>>>,
+        resources: Vec<IoResource>,
+        irq: Option<IrqResource>,
+    ) -> DeviceDescriptor {
+        DeviceDescriptor::new(id, name, dev.clone(), parent_bus, resources, irq)
+    }
+
+    // Allocate IO and instance id resources.
+    // Return a Result wrapper of instance id.
+    fn allocate_resources(&mut self, resources: &mut Vec<IoResource>) -> Result<(u32)> {
+        let id = self
+            .resource
+            .lock()
+            .expect("failed to acquire lock.")
+            .allocate_instance_id()
+            .map_err(Error::InstanceIdAllocate)?;
+
+        for res in resources.iter_mut() {
+            match res.res_type {
+                IoType::Pio => {
+                    if res.addr.is_none() {
+                        return Err(Error::NonePIORequest);
+                    }
+                    res.addr = Some(
+                        self.resource
+                            .lock()
+                            .expect("failed to acquire lock")
+                            .allocate_io_addresses(res.addr.unwrap(), res.size)
+                            .map_err(Error::AddressAllocate)?,
+                    );
+                }
+                IoType::PhysicalMmio | IoType::Mmio => {
+                    res.addr = Some(
+                        self.resource
+                            .lock()
+                            .expect("failed to acquire lock")
+                            .allocate_mmio_addresses(res.addr, res.size)
+                            .map_err(Error::AddressAllocate)?,
+                    )
+                }
+            }
+        }
+        Ok(id)
+    }
+
+    fn free_resources(&mut self, resources: &[IoResource], id: u32) {
```

Comment:
let res = self
                    .resource
                    .lock()
                    .expect("failed to acquire lock");
so we don't to lock/unlock for every resource entry.

Reply:
Once we protect DeviceManager in VMM, and keep SystemAllocator members thread safe, the resources are protected. Do I understand correctly for your comments? (The codes with the last updated patch)

Reply:
I means just acquire the lock once at entries, instead of acquire/release for each accessed fields.